modules/rostests/apitests/ntdll/NtQueryInformationProcess.c

```diff
@@ -323,6 +325,114 @@ Test_ProcessWx86Information(void)
+        }
+
+        /* We query an invalid buffer address */
+        Status = NtQueryInformationProcess(NtCurrentProcess(),
+                                           InfoClass,
+                                           (PVOID)(ULONG_PTR)PsProcessInfoClass[InfoClass].AlignmentQUERY,
+                                           PsProcessInfoClass[InfoClass].RequiredSizeQUERY,
+                                           NULL);
+
+        /* Ensure that we do not fail on classes that don't belong to NtQueryInformationProcess */
+        if (!(PsProcessInfoClass[InfoClass].Flags & ICIF_QUERY) && Status != STATUS_DATATYPE_MISALIGNMENT)
+        {
+            ok(Status == STATUS_INVALID_INFO_CLASS, "STATUS_INVALID_INFO_CLASS status expected, got 0x%lx! (Information class -> %lu)\n", Status, InfoClass);
+        }
+        else if (InfoClass == ProcessWorkingSetWatch || InfoClass == ProcessHandleTracing)
+        {
+            /* Ensure that we test specific special status values against specific classes */
+            ok(Status == SpecialStatus, "0x%lx special status expected, got 0x%lx! (Information class -> %lu)\n", SpecialStatus, Status, InfoClass);
+        }
+        else
+        {
+            ok(Status == STATUS_ACCESS_VIOLATION, "STATUS_ACCESS_VIOLATION status expected, got 0x%lx! (Information class -> %lu)\n", Status, InfoClass);
+        }
+
+        /* The information length is wrong */
+        Status = NtQueryInformationProcess(NtCurrentProcess(),
+                                           InfoClass,
+                                           (PVOID)(ULONG_PTR)PsProcessInfoClass[InfoClass].AlignmentQUERY,
+                                           PsProcessInfoClass[InfoClass].RequiredSizeQUERY - 1,
+                                           NULL);
+
+        /*
+         * Ensure that we do not fail on classes that don't belong to NtQueryInformationProcess.
+         * Although there are also classes that aren't originally implemented in Windows Server 2003
+         * but fail with STATUS_ACCESS_VIOLATION regardless...
+        */
+        if (!(PsProcessInfoClass[InfoClass].Flags & ICIF_QUERY) && Status != STATUS_DATATYPE_MISALIGNMENT)
+        {
+            ok(Status == STATUS_INVALID_INFO_CLASS || Status == STATUS_ACCESS_VIOLATION,
+               "STATUS_INVALID_INFO_CLASS or STATUS_ACCESS_VIOLATION status expected, got 0x%lx! (Information class -> %lu)\n", Status, InfoClass);
+        }
+        else if (InfoClass == ProcessWorkingSetWatch || InfoClass == ProcessHandleTracing)
+        {
+            /* Ensure that we test specific special status values against specific classes */
+            ok(Status == SpecialStatus, "0x%lx special status expected, got 0x%lx! (Information class -> %lu)\n", SpecialStatus, Status, InfoClass);
+        }
+        else
+        {
+            ok(Status == STATUS_INFO_LENGTH_MISMATCH || Status == STATUS_ACCESS_VIOLATION,
```

Comment:
Isn't STATUS_ACCESS_VIOLATION here a sign that our `RequiredSizeQUERY` value is wrong (since the function will "succeed" with a smaller size)? Or is this just XP compat?

Reply:
Server 2003's `NtQueryInformationProcess` (as other functions) has a weird mechanism into probing the class data such as buffer alignment, size, etc. first and then proceeding further into a kind of switch case, which class exists for Server 2003 and not.

For classes that aren't implemented in Windows Server 2003, let's take `ProcessAffinityUpdateMode ` for example, `NtQueryInformationProcess` indiscriminately probes the size even for classes that don't literally exist in Windows Server 2003. I admit I didn't comment the testcase code enough so everyone would understand the motive behind such tests.

So to answer your question, yes, STATUS_ACCESS_VIOLATION here indicates `RequiredSizeQUERY` value is wrong but only for classes that don't exist in Windows Server 2003.

Reply:
Classes that don't exist in Server 2003 shouldn't have the `ICIF_QUERY` bit set, and therefore should go into the first if branch, right? So maybe it means our table is wrong after all?